src/rules/add-claim-based-on-relationships.js

```diff
@@ -0,0 +1,121 @@
+* - GROUPS4AUTH0_CUSTOMCLAIM: A namespaced custom claim name of your choice. The name must be a URL.
+* - GROUPS4AUTH0_AUDIENCE: The audience in the machine-to-machine token.
+*/
+function addClaimBasedOnRelationships(user, context, callback) {
+    const fetch = require("node-fetch");
+    const { URLSearchParams } = require('url');
+    const jwt = require('jsonwebtoken');
+
+    const getM2MToken = () => {
+        if (global.m2mToken && global.m2mTokenExpiryInMillis > new Date().getTime() + 60000) {
+            return Promise.resolve(global.m2mToken);
+        } else {
+            return fetch(`https://${auth0.domain}/oauth/token`, {
+                method: 'POST',
+                body: new URLSearchParams({
+                    grant_type: 'client_credentials',
+                    client_id: configuration.GROUPS4AUTH0_CLIENTID,
+                    client_secret: configuration.GROUPS4AUTH0_CLIENTSECRET,
+                    audience: configuration.GROUPS4AUTH0_AUDIENCE,
+                    scope: 'pg:tenant:admin'
+                })
+            })
+                .then(response => {
+                    if (!response.ok) {
+                        return response.text().then((error) => {
+                            console.error("Failed to obtain m2m token");
+                            throw Error(error);
+                        });
+                    } else {
+                        return response.json();
+                    }
+                })
+                .then(({ access_token }) => {
+                    global.m2mToken = access_token;
+                    global.m2mTokenExpiryInMillis = jwt.decode(access_token).exp * 1000;
+                    return access_token;
+                });
+        }
+    };
+
+    const callPrivateGroups = (accessToken, path) => {
+        const url = `${configuration.GROUPS4AUTH0_BASEURL}/${configuration.GROUPS4AUTH0_TENANT}/${path}`;
+        return fetch(url, {
+            method: 'GET',
+            headers: {
+                "Authorization": "Bearer " + accessToken,
+                "Content-Type": "application/json"
+            }
+        })
+            .then(response => {
+                if (response.status === 404) {
+                    return [];
+                } else if (!response.ok) {
+                    return response.text().then((error) => {
+                        console.error("Failed to call private groups v2", url);
+                        throw Error(error);
+                    });
+                } else {
+                    return response.json();
+                }
+            });
+    };
+
+    const getRelationships = (accessToken) => {
+        return callPrivateGroups(accessToken, `actors/user/${user.user_id}/relationships`);
+    };
+
+    const addClaimToToken = (privateGroupsResponse) => {
+        context.accessToken[configuration.GROUPS4AUTH0_CUSTOMCLAIM] = privateGroupsResponse.map(relationship => ({
```

Comment:
Might be good to have a default for `configuration.GROUPS4AUTH0_CUSTOMCLAIM`

Reply:
Also, did you test that this array of objects comes through as expected? Just want to make sure, some of the claims expect a specific format (`scope`, for example, is an array of strings).

Reply:
Added the default and added a unit test